explainer.md

```diff
@@ -36,193 +97,126 @@ self.addEventListener('launch', event => {
 
-This section is a rough draft of "spec-ish" language, without being too picky
-about getting things "right". See also the [polyfill source
-code](demos/polyfill.js), which roughly implements this logic.
+User agents can give feedback to users when a site is handling a `launch` event to signify that the app is loading. User agents have a lot of flexibility to experiment here but some suggestions on what could be done if the app doesn't show some UI after a small delay (e.g. 1-2 seconds):
+- Show a splash screen indicating the app is launching
+- Show an entry for the app in the taskbar/dock/shelf indicating it's loading
+- Focus a previously opened window in the scope of the app
 
-## LaunchEvent
+If the app doesn't show UI after a long delay (e.g. 10 seconds), the user agent could:
+- Kill the `launch` event handler and show an error message indicating the app couldn't launch
+- If apps behave badly on a repetitive basis, don't allow it to handle `launch` events (fallback to opening URLs directly in their default context)
+
+#### Solution 2: `launch` event on Window rather than Service Worker
+
+Another solution to this problem that has been proposed is to handle `launch` events in windows rather than in the Service Worker. If an existing window in the scope of the web app is opened and had registered a `launch` handler, the event would be sent there and the window focused immediately. If no existing window is opened, a new window would automatically be opened, navigated to the target URL and the event would then be sent (after the `onload` event had fired).
+
+This proposal restricts what would be possible with `launch` events, and could also result in a poor user experience:
+- It would not be possible to allow websites to only show a notification from the `launch` event and not display any window
+- If an existing window is open and the app wants to open a new window, a flicker of the existing window will be shown as it is focused prior triggering the event
+- In terms of ergonomics, developers must be careful to register `launch` event handlers on all windows and to co-ordinate between them. Often this will involve going through the service worker anyway.
+
+### Whether `launch` events should *only* be triggered by navigations
+
+In the proposal described so far, all `launch` events are triggered by navigations to URLs. However, it's an open question whether this would be a pre-requisite for future `launch` events. The File Handlers API is introducing the ability for web apps to register themselves as handlers for certain file types. When designing this API there is a choice:
+1. Pass file handles to the app via the `launch` event in the service worker
+2. Invent a new way of passing file handles to windows. For example, the browser could launch a URL and pass blob URLs as query parameters, e.g. https://example.com/file_handler?file1=blob://abcdefg&file2=blob://tuvwxyz.
```

Comment:
Did you manage to find out if we can persist stuff we wrote to the blob?

Reply:
There is no existing way to write to a blob:// url but I don't see any glaring reason why something like this approach wouldn't work. I'll leave it to experts to give their perspective :)

Reply:
Hopefully we can just reuse whatever it is the writable files people come up with for persisting file handles.